### PR TITLE
Add team export/import scripts [SCI-1849]

### PR DESCRIPTION
**Pull request by okriuchykhin**
https://biosistemika.atlassian.net/browse/SCI-1849

### CODE REVIEW

File: app/services/team_exporter.rb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+      puts 'Done!'
+    end
+  end
+
+  private
+
+  def copy_files(assets, attachment_name, dir_name)
+    assets.flatten.each do |a|
+      next unless a.public_send(attachment_name).present?
+      @asset_counter += 1 if attachment_name == :file
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_exporter.rb]: done

File: lib/tasks/data.rake [hunk 2]
```diff
@@ -74,4 +74,20 @@ namespace :data do
   desc "Remove temporary and obsolete data"
   task clean: [:environment, :clean_temp_files, :clean_unconfirmed_users]
 
+  desc 'Export team to directory'
+  task :team_export, [:team_id] => [:environment] do |_, args|
+    Rails.logger.info(
+      "Exporting team with ID:#{args[:team_id]} to directory in tmp"
+    )
+    te = TeamExporter.new(Team.find_by_id(args[:team_id]))
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply lib/tasks/data.rake]: done

File: app/services/team_exporter.rb [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      custom_fields: team.custom_fields,
+      repositories: team.repositories.map { |r| repository(r) },
+      tiny_mce_assets: team.tiny_mce_assets,
+      protocols: team.protocols.where(my_module: nil).map { |pr| protocol(pr) },
+      protocol_keywords: team.protocol_keywords,
+      projects: team.projects.map { |p| project(p) }
+    }
+  end
+
+  def notification_json(notification)
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_exporter.rb]: done

File: app/services/team_exporter.rb [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    # Looks like Devise doesn't export some fields to JSON, so add it manually
+    user_json['encrypted_password'] = user.encrypted_password
+    user_json['confirmed_at'] = user.confirmed_at
+    user_json['sign_in_count'] = user.sign_in_count
+    user_json['last_sign_in_at'] = user.last_sign_in_at
+    user_json['last_sign_in_ip'] = user.last_sign_in_ip
+    copy_files([user], :avatar, File.join(@dir_to_export, 'avatars'))
+    {
+      user: user_json,
+      user_notifications: user.user_notifications,
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_exporter.rb]: Notifications are not scoped by the team actually. And in most cases they are assigned to project users

File: app/services/team_importer.rb [hunk 5]
```diff
@@ -0,0 +1,860 @@
...
+
+  def create_users(users_json, team)
+    puts 'Creating users...'
+    users_json.each do |user_json|
+      user = User.new(user_json['user'])
+      orig_user_id = user.id
+      user.id = nil
+      user.password = user_json['user']['encrypted_password']
+      user.current_team_id = team.id
+      user.invited_by_id = nil
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: done

File: app/services/team_importer.rb [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+      repository_row.created_by_id = find_user(repository_row.created_by_id)
+      repository_row.last_modified_by_id =
+        find_user(repository_row.last_modified_by_id)
+      repository_row.save!
+      @repository_row_mappings[orig_rep_row_id] = repository_row.id
+      @repository_row_counter += 1
+      repository_row_json['my_module_repository_rows'].each do |mm_rep_row_json|
+        mm_rep_row = MyModuleRepositoryRow.new(mm_rep_row_json)
+        mm_rep_row.id = nil
+        mm_rep_row.repository_row = repository_row
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: fixed

File: app/services/team_importer.rb [hunk 7]
```diff
@@ -0,0 +1,860 @@
...
+      orig_step_id = step.id
+      step.id = nil
+      step.user_id = find_user(step.user_id)
+      step.last_modified_by_id = find_user(step.last_modified_by_id)
+      step.protocol_id = protocol.id
+      step.save!
+      @step_mappings[orig_step_id] = step.id
+      @step_counter += 1
+
+      step_json['step_comments'].each do |step_comment_json|
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: yes, it works

File: app/services/team_importer.rb [hunk 8]
```diff
@@ -0,0 +1,860 @@
...
+        my_module_group = MyModuleGroup.new(my_module_group_json)
+        orig_module_group_id = my_module_group.id
+        my_module_group.id = nil
+        my_module_group.experiment = experiment
+        my_module_group.created_by_id =
+          find_user(my_module_group.created_by_id)
+        my_module_group.save!
+        @my_module_group_mappings[orig_module_group_id] = my_module_group.id
+      end
+      experiment.delay.generate_workflow_img
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: If transaction fails - no delayed_job records will be stored in DB

File: app/services/team_importer.rb [hunk 9]
```diff
@@ -0,0 +1,860 @@
...
+        sample_module = SampleMyModule.new(sample_module_json)
+        sample_module.id = nil
+        sample_module.sample_id =
+          @sample_mappings[sample_module.sample_id]
+        sample_module.my_module = my_module
+        sample_module.assigned_by_id =
+          find_user(sample_module.assigned_by_id)
+        sample_module.save!
+      end
+
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: Just moved repository creation later in the code

File: app/services/team_importer.rb [hunk 10]
```diff
@@ -0,0 +1,860 @@
...
+    connections_json.each do |connection_json|
+      connection = Connection.new(connection_json)
+      connection.id = nil
+      connection.input_id = @my_module_mappings[connection.input_id]
+      connection.output_id = @my_module_mappings[connection.output_id]
+      connection.save!
+    end
+  end
+
+  def create_activity(activities_json)
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: done

File: app/services/team_importer.rb [hunk 11]
```diff
@@ -0,0 +1,860 @@
...
+    repository_cells_json.each do |repository_cell_json|
+      repository_cell =
+        RepositoryCell.new(repository_cell_json['repository_cell'])
+      repository_cell.id = nil
+      repository_cell.repository_column_id =
+        @repository_column_mappings[repository_cell.repository_column_id]
+      repository_cell.repository_row_id =
+        @repository_row_mappings[repository_cell.repository_row_id]
+      repository_value =
+        RepositoryTextValue.new(repository_cell_json['repository_value'])
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: probably yes

File: app/services/team_importer.rb [hunk 12]
```diff
@@ -0,0 +1,860 @@
...
+        end
+      end
+
+      team_json['projects'].each do |project_json|
+        create_activity(project_json['activities'])
+        create_reports(project_json['reports'])
+        project_json['experiments'].each do |experiment_json|
+          experiment_json['my_modules'].each do |my_module_json|
+            create_task_connections(my_module_json['inputs'])
+            create_task_connections(my_module_json['outputs'])
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: We need to do it only after all tasks are created.
Ok, I think we can process only outputs assuming that connection can't exist without input.

File: lib/tasks/data.rake (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply lib/tasks/data.rake]: :+1:

File: app/services/team_exporter.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_exporter.rb]: :+1:

File: app/services/team_exporter.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_exporter.rb]: :+1:

File: app/services/team_exporter.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_exporter.rb]: :+1:

File: app/services/team_importer.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: Ah, glad to hear that! So this is something we got covered :+1:

File: app/services/team_importer.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb [hunk 13]
```diff
@@ -0,0 +1,860 @@
...
+      Protocol.set_callback(:save, :after, :update_linked_children)
+    end
+  end
+
+  private
+
+  def update_smart_annotations(team)
+    team.projects.each do |pr|
+      pr.project_comments.each do |comment|
+        comment.save! if update_annotation(comment.message)
```
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: :+1:

File: app/services/team_importer.rb [hunk 14]
```diff
@@ -0,0 +1,860 @@
...
+        result_text.save!
+        @result_text_mappings[orig_result_text_id] = result_text.id
+      end
+
+      result.save!
+      @result_mappings[orig_result_id] = result.id
+      @result_counter += 1
+
+      result_json['result_comments'].each do |result_comment_json|
+        result_comment = ResultComment.new(result_comment_json)
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/services/team_importer.rb]: yes

**Comment by Ducz0r** [comment]
One additional thing: we were just discussing this a day ago with people from legal - can you update the migration script so it will only migrate the "active" users (the users who are "pending", e.g. haven't confirmed their invitation/email yet, should not be migrated, so can probably be ignored when preparing the exported JSON file, as well as their avatar).

File: app/services/team_importer.rb [hunk 15]
```diff
@@ -0,0 +1,860 @@
...
+      end
+      if tiny_mce_asset.result_text_id.present?
+        result_text = ResultText.find_by_id(tiny_mce_asset.result_text_id)
+        result_text.text.sub!("[~tiny_mce_id:#{orig_tmce_id}]",
+                              "[~tiny_mce_id:#{tiny_mce_asset.id}]")
+      end
+    end
+  end
+
+  def create_users(users_json, team)
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/services/team_importer.rb]: @okriuchykhin , @Ducz0r have you decide how to handle this scenario?

[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: Not yet. But I'll open a discussion on Slack right now.

File: app/services/team_importer.rb [hunk 16]
```diff
@@ -0,0 +1,878 @@
...
+
+  def import_from_dir(import_dir)
+    MyModule.skip_callback(:create, :before, :create_blank_protocol)
+    Activity.skip_callback(:create, :after, :generate_notification)
+    Protocol.skip_callback(:save, :after, :update_linked_children)
+    @import_dir = import_dir
+    team_json = JSON.parse(File.read("#{@import_dir}/team_export.json"))
+    team = Team.new(team_json['team'].slice(*Team.column_names))
+    team.id = nil
+    team.transaction(isolation: :serializable) do
```
Comment (ZmagoD) [inline app/services/team_importer.rb]: when we create a team how we handle the billing plan and subscriptions,....

File: app/services/team_importer.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (Ducz0r) [reply app/services/team_importer.rb]: At the moment import will be cancelled if email is already in use as we have uniqueness constrain on user - so we are dealing with this case in a way.

**Comment by ZmagoD** [comment]
And the problem is not the AWS S3 key. Is working fine for uploading files in sciNote :thinking:

**Comment by okriuchykhin** [comment]
So probably you have corrupted database where some Assets are without actual files

**Pull request merged**